### PR TITLE
Change to namespace package

### PR DESCRIPTION
**Pull request by sebkuf**
It should change the ``intelmq/bots`` folder to a namespace package. Now it should be easier to add new Bots. The ``intelmqctl list bots`` functionality now also works with namespaces. Added a documentation how to add a new bot.
Implements #2350

### CODE REVIEW

File: intelmq/lib/utils.py [hunk 1]
```diff
@@ -860,13 +862,22 @@ def list_all_bots() -> dict:
...
+    ns_pkg = intelmq.bots
 
-    botfiles = [botfile for botfile in pathlib.Path(base_path).glob('**/*.py') if botfile.is_file() and botfile.name != '__init__.py']
-    for file in botfiles:
-        file = Path(file.as_posix().replace(base_path, 'intelmq/bots'))
+    def iter_ns(pkg):
+        botmodules = []
+        for path, name, is_pkg in iter_modules(pkg.__path__, pkg.__name__ + '.'):
+            if is_pkg:
+                botmodules = botmodules + iter_ns(importlib.import_module(name))
```
Comment (sebix) [inline intelmq/lib/utils.py:871]: This line causes the current unit test failures (import of a SyntaxError-throwing-bot fails with a SyntaxError, which is not handled correctly here). An example of catching the exception is nine lines below.

Comment (sebix) [inline intelmq/lib/utils.py:871]: This line causes the current unit test failures (import of a SyntaxError-throwing-bot fails with a SyntaxError, which is not handled correctly here). An example of catching the exception is nine lines below.

File: intelmq/bots/__init__.py [hunk 2]
```diff
@@ -0,0 +1 @@
+__path__ = __import__('pkgutil').extend_path(__path__, __name__)
```
Comment (sebix) [inline intelmq/bots/__init__.py:1]: trailing newlines missing here and in the other similar files

File: aTestNamespace/intelmqTest.egg-info/PKG-INFO [hunk 3]
```diff
@@ -0,0 +1,33 @@
+Metadata-Version: 2.1
```
Comment (sebix) [inline aTestNamespace/intelmqTest.egg-info/PKG-INFO:1]: I think these files should be part of the repository, should they?

File: contrib/examples/example_add_bot/setup.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
+from setuptools import find_packages, setup
+
+setup(
+    name='intelmqExampleAddBot',
+    version=3.1,  # noqa: F821
```
Comment (sebix) [inline contrib/examples/example_add_bot/setup.py:5]: should be a string AFAIK. And let's use semantic versioning
```suggestion
    version='3.1.0',
```

File: contrib/examples/example_add_bot/setup.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    maintainer='Sebastian Wagner',
+    maintainer_email='intelmq-dev@lists.cert.at',
```
Comment (sebix) [inline contrib/examples/example_add_bot/setup.py:7]: I don't want to be responsible for code which I never wrote and never saw =)
```suggestion
    maintainer='Your Name',
    maintainer_email='you@example.com',
```

File: contrib/examples/example_add_bot/setup.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    python_requires='>=3.7',
```
Comment (sebix) [inline contrib/examples/example_add_bot/setup.py:8]:
```suggestion
```
intelmq itself requires python >= 3.7 currently, that may not be true for bots. as the bots require the core library anyway, the requirement is active too without listing it here

File: docs/dev/guide.rst [hunk 7]
```diff
@@ -891,6 +891,50 @@ The databases `<` 10 are reserved for the IntelMQ core:
...
+   ├── setup.py
+   └── intelmq
+        ├── __init__.py
+        └── bots
+             └── collectors
+                  └── your_bot
+                       └── your_bot.py
+
+The ``additional_bots`` folder needs to be defined e.g. with a ``setup.py``. Only the intelmq folder contains a ``__init__.py`` 
+file and it must contain only this line. Any additional code in ``__init__.py`` after this line will be inaccessible.
```
Comment (sebix) [inline docs/dev/guide.rst:912]:
```suggestion
file, which must have only the following line. Any additional code in ``__init__.py`` after this line will be inaccessible.
```

File: intelmq/lib/utils.py [hunk 8]
```diff
@@ -860,13 +862,22 @@ def list_all_bots() -> dict:
     from intelmq.lib.bot import Bot  # noqa: prevents circular import
     bot_parameters = dir(Bot)
 
-    base_path = resource_filename('intelmq', 'bots')
+    ns_pkg = intelmq.bots
 
-    botfiles = [botfile for botfile in pathlib.Path(base_path).glob('**/*.py') if botfile.is_file() and botfile.name != '__init__.py']
-    for file in botfiles:
-        file = Path(file.as_posix().replace(base_path, 'intelmq/bots'))
+    def iter_ns(pkg):
```
Comment (sebix) [inline intelmq/lib/utils.py:867]: Could be move the function to the outer scope?

**Review by sebix — COMMENTED** [review]
Thank you for your work so far. Would be very cool if we can get this feature done :)

**Comment by kamil-certat** [comment]
I'll take over the change and create a new PR as sebkuf is no longer available.

**Pull request closed without merging**